File: Exercices/Visitor/word_visitor/browsable_word_search_tree.rb
require './word_search_tree'

class BrowsableWordSearchTree < WordSearchTree

  class Node < WordSearchTree::Node
    def browse(visitor, word)
      visitor.visit(word) if word?
      
      @children.each_with_index do |child, index|
        next unless child
        char = (BASE+index).chr
        child.browse(visitor, word + char)
      end
    end
  end

  def initialize
    @root = Node.new
  end

  def browse(visitor)
    @root.browse(visitor, '')
  end
end

File: Exercices/Visitor/java/src/BrowsableWordSearchTree.java
public class BrowsableWordSearchTree extends WordSearchTree implements WordsBrowsable {
    class Node extends WordSearchTree.Node {
        void browse(WordsVisitor visitor, String partialWord) {
            if (isWord) {
                visitor.visit(partialWord);
            }
            
            for (int childIndex = 0; childIndex < 26; childIndex++) {
                if (children[childIndex] != null) {
                    ((Node)children[childIndex]).browse(visitor, partialWord + (char)(BASE+childIndex));
                }
            }
        }
    }
    
    @Override
    public void browse(WordsVisitor visitor)
    {
        ((Node)root).browse(visitor, "");
    }

    @Override
    protected Node createNode() {
        return new Node();
    }
    
}
